From bee9f5e5eff62eb6a3f52127f1dc6de2bbffe2b9 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Wed, 4 Nov 2015 20:17:01 -0800
Subject: [PATCH 0855/1767] UBUNTU: SAUCE: apparmor: Fix: refcount leak in
 aa_label_merge

if aa_label_alloc() fails the refs taken on a and b are leaked.

Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index b40c877..d011c14 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -1099,7 +1099,7 @@ struct aa_label *aa_label_merge(struct aa_label *a, struct aa_label *b,
 		 */
 		new = aa_label_alloc(a->size + b->size, gfp);
 		if (!new)
-			return NULL;
+			goto out;
 
 		write_lock_irqsave(&ls->lock, flags);
 		label = __label_merge_insert(ls, new, a, b);
@@ -1110,6 +1110,7 @@ struct aa_label *aa_label_merge(struct aa_label *a, struct aa_label *b,
 			new = NULL;
 		}
 		aa_put_label(new);
+	out:
 		aa_put_label(a);
 		aa_put_label(b);
 	}
-- 
1.9.3

